### PR TITLE
fix(linter): flag banned external imports reached through internal projects

### LINKED ISSUE #36519: @nx/eslint-plugin: checkNestedExternalImports never flags transitive bannedExternalImports reached via an internal project
### Current Behavior

With `checkNestedExternalImports: true` and a `bannedExternalImports` constraint, importing an **internal** project whose *transitive* dependencies include the banned external package produces **no** violation. Only *direct* imports of the banned package are caught (by the separate, non-nested path), so `checkNestedExternalImports` effectively does nothing.

**Root cause.** In `hasBannedDependencies` → `isConstraintBanningProject(externalProject, constraint, imp)` (`packages/eslint-plugin/src/utils/runtime-lint-utils.ts`) the first guard is:

    const { packageName } = externalProject.data;
    if (imp !== packageName && !imp.startsWith(`${packageName}/`)) {
      return false;
    }

`packageName` is the **transitive external** node (e.g. `lodash`), but `imp` is the **original import specifier** forwarded from the rule. In the only path that reaches the nested check (`enforce-module-boundaries.ts`: the `targetProject.type === 'npm'` branch `return`s earlier, so the nested block runs **only for internal targets**), `imp` is the *internal* project specifier (e.g. `@org/middle`). It can never equal the transitive external `packageName`, so the guard always returns `false` and `hasBannedDependencies` returns `[]`.

The unit test that appears to cover this (`runtime-lint-utils.spec.ts`) calls `hasBannedDependencies(deps, graph, constraint, 'react-native')` — passing `imp` = the *external* package name, a value the rule never supplies on the nested path. There is **no** end-to-end test for `checkNestedExternalImports` in `enforce-module-boundaries.spec.ts`. The same code is present on `master`.

### Expected Behavior

`checkNestedExternalImports: true` should flag when a constrained project imports an internal project that (transitively) depends on a `bannedExternalImports` package, emitting `nestedBannedExternalImportsViolation` ("… Nested import found at `{childProjectName}`"). This matches the feature's original intent in #11000.

### GitHub Repo

_No response_

### Steps to Reproduce

```bash
# 1. Create workspace
npx create-nx-workspace@23.1.0 nxrepro --preset=ts --nxCloud=skip --no-interactive --packageManager=npm
cd nxrepro

# 2. Generate two libs
npx nx g @nx/js:lib packages/middle   --no-interactive
npx nx g @nx/js:lib packages/consumer --no-interactive

# 3. Add ESLint tooling (no lint target/config exists by default)
npx nx add @nx/eslint
npm i -D @nx/eslint-plugin@23.1.0 typescript-eslint
npm i lodash
```

Edit `packages/middle/src/lib/middle.ts`:
```ts
import { camelCase } from 'lodash';
export function middle() {
  return camelCase('hello world');
}
```

Edit `packages/consumer/src/lib/consumer.ts`:
```ts
import { middle } from '@nxrepro/middle';
export function consumer() {
  return middle();
}
```

Add `lodash` as a real dependency of `middle` — `packages/middle/package.json`:
```jsonc
{
  // ...
  "dependencies": {
    "tslib": "^2.3.0",
    "lodash": "^4.17.21"
  }
}
```

Tag `consumer` — `packages/consumer/package.json`:
```jsonc
{
  // ...
  "nx": {
    "tags": ["scope:consumer"]
  }
}
```

Create root `eslint.config.mjs`:
```js
import nx from '@nx/eslint-plugin';
import tseslint from 'typescript-eslint';

export default [
  {
    files: ['**/*.ts', '**/*.tsx', '**/*.js', '**/*.jsx'],
    languageOptions: { parser: tseslint.parser },
    plugins: { '@nx': nx },
    rules: {
      '@nx/enforce-module-boundaries': ['error', {
        checkNestedExternalImports: true,
        depConstraints: [
          { sourceTag: 'scope:consumer', bannedExternalImports: ['lodash'] },
        ],
      }],
    },
  },
];
```

```bash
# 4. Run lint
npx nx reset
npx nx lint @nxrepro/consumer
```

### Actual behavior
```
> nx run @nxrepro/consumer:lint
> eslint .

 NX   Successfully ran target lint for project @nxrepro/consumer
```
No violation is reported.

### Nx Report

```shell
Node           : 22.22.2
OS             : linux-x64
Native Target  : x86_64-linux
yarn           : 1.22.22
daemon         : Available

nx                 : 23.1.0
@nx/js             : 23.1.0
@nx/workspace      : 23.1.0
@nx/devkit         : 23.1.0
@nx/eslint-plugin  : 23.1.0
typescript         : 5.9.3
```

### Failure Logs

```shell

```

### Package Manager Version

_No response_

### Operating System

- [ ] macOS
- [x] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

- The feature originally worked in #11000, where `hasBannedDependencies` was called with 3 args (no `imp`). The `imp` parameter + guard were added later (to support subpath matching on the direct path) and broke the nested path.
- **Suggested fix:** on the nested path, match the transitive external node's own `packageName` against `bannedExternalImports` (apply the `imp`-based subpath guard only on the direct import path), or don't forward `imp` into `hasBannedDependencies`.

### PR DESCRIPTION
**Pull request by leosvelperez**
## Current Behavior

With `checkNestedExternalImports: true`, importing an internal project never reports a violation for that project's transitive external dependencies:

- With a `bannedExternalImports` constraint, a child project depending on the banned package produces no violation. Only direct imports of the package are flagged, so the nested check does nothing.
- With an `allowedExternalImports` constraint and no ban list, the nested check is skipped entirely. A child project depending on a package outside the allowed list is never reported.

## Expected Behavior

Importing an internal project whose transitive external dependencies violate a `bannedExternalImports` or `allowedExternalImports` constraint reports a violation. The error names the violating package and the child project where it was found. Each package is reported once per child project. The allowed list is exclusive on the nested path, as it already is for direct imports.

## Related Issue(s)

Fixes #36519

## Implementation Notes

- The subpath-matching guard #17755 added to `isConstraintBanningProject` compares the original import specifier against the external package name. On the nested path the specifier names the imported internal project while the compared package is the transitive external, so the guard never matched. Nested matching now uses each transitive package's own name. The direct import path keeps the specifier-based subpath matching.
- Nested matching stays package-granular, same as before #17755, since the project graph records external dependencies at the package level. A `lodash/fp` ban does not flag a child depending on `lodash`. The rule options docs state this.
- The nested check ran only when a constraint had a ban list, a gate written before `allowedExternalImports` existed. The gate accepts an allowed list too, including an empty one, which bans every external package.
- `findTransitiveExternalDependencies` keeps one dependency edge per project and package. Several edges to the same package (static and dynamic, or several resolved versions) produce a single violation instead of identical repeats.
- Rule-level tests cover both directions for banned, allowed-only, mixed and empty-allowed constraints, plus multi-hop chains and the duplicate-edge cases.

---
<p><a href="https://app.trypolygraph.com/orgs/6a061dcb561c062131116eca/sessions/gh-36519-1f4c8a12">View Polygraph session ↗</a></p>

### CODE REVIEW

**Review by AgentEnder — CHANGES_REQUESTED** [review]
Confirmed the fix does what it says. I ran your new specs against master's version of the two source files, and exactly the four tests that assert a nested violation fail there. All 118 pass at HEAD. The feature was fully inert before this.

One thing I'd want changed before it lands.

## The violation message never names the package that violated

`nestedBannedExternalImportsViolation` interpolates `imp`, which on this path is the internal import. The transitive package that actually matched the constraint never appears.

That was harmless while the branch was unreachable. Now that `allowedExternalImports` reaches it, the error blames an import that is itself permitted:

```
A project tagged with "api" is not allowed to import "@mycompany/impl". Nested import found at implName
```

Same finding, second half: `hasBannedDependencies` returns one tuple per dependency edge and the rule reports one error per tuple. With your own fixture, `implName` depending on `npm-package` (static), `npm-awesome-package` (static) and `npm-package` (dynamic) produces 3 byte-identical reports for a single `import`. `findTransitiveExternalDependencies` doesn't dedupe, despite its jsdoc promising unique deps.

Adding `target.data.packageName` to the report data and the message template covers the first half. Keying the filtered tuples by `${dep.source}|${dep.target}` covers the second.

## Maintainer calls

Three decisions worth naming rather than inheriting. All three read as deliberate to me and I'd land them as they are:

- The gate widening to `allowedExternalImports` goes past #36519. It's the coherent choice, since the allow branch already ran nested for mixed constraints and fixing only the ban list would leave `{allowed: [...]}` silently unenforced. But anyone already running `checkNestedExternalImports` next to an allow list goes from zero nested errors to one per transitive package outside that list.
- `allowedExternalImports: []` now bans every transitive external, because the gate tests truthiness. That matches the direct path.
- Nested matching is package-granular, so `['lodash/fp']` won't catch a nested `lodash`. The options table doesn't mention the asymmetry.

## Smaller things

- Every new test has the banned package exactly one hop away. Two hops works, and names the right child, but nothing pins it.
- `runtime-lint-utils.spec.ts:349` is now identical in inputs and expectation to the test at :314.
- `imp` means "import specifier" on the direct path and "package name" on the nested one. A rename would carry that without a comment.

**Comment by leosvelperez** [comment]
Thanks for the thorough pass @AgentEnder, and for running the specs against master to confirm the fix.

Good catch on the message. It now names the package: `... is not allowed to import "@mycompany/impl". Nested import of "npm-package" found at implName`.

For the duplicates I went with deduping in `findTransitiveExternalDependencies`, but keyed by source project + package name rather than node name, so two versioned nodes of the same package (`npm:pkg`, `npm:pkg@2.0.0`) also collapse into one report. Both cases have tests now.

Also added a two-hop test, removed the lint-utils test that had become identical to its sibling, renamed the `isConstraintBanningProject` param to `importSpecifier`, and documented the package-granular matching in the options table.

The three maintainer calls stay as you read them: all deliberate.

File: astro-docs/src/content/docs/kb/enforce-module-boundaries.mdoc [hunk 1]
```diff
@@ -73,16 +73,16 @@ export default [
...
+| checkDynamicDependenciesExceptions | _Array<string>_           | _[]_    | List of imports that should be skipped for `Imports of lazy-loaded libraries forbidden` checks. E.g. `['@myorg/lazy-project/component/*', '@myorg/other-project']`                                                                                                                                                                                                 |
+| checkNestedExternalImports         | _boolean_                 | _false_ | Enable to also apply the `bannedExternalImports` and `allowedExternalImports` constraints to the external packages that imported projects depend on, transitively. The nested check matches whole package names, so subpath entries such as `lodash/fp` only apply to direct imports. Check [Dependency constraints](#dependency-constraints) for more information |
```
Comment (AgentEnder) [inline astro-docs/src/content/docs/kb/enforce-module-boundaries.mdoc]: This holds for `bannedExternalImports`, but not for `allowedExternalImports`, which the same cell says the flag now applies to. A subpath entry there is not inert on the nested path: it is what causes the violation.

```
allowedExternalImports: ['lodash/fp'], child depends on lodash
  -> A project tagged with "x" is not allowed to import "@org/middle".
     Nested import of "lodash" found at middle

bannedExternalImports: ['lodash/fp'], same child
  -> no report
```

The whitelist arm evaluates `!importSpecifier.startsWith(packageName) || !mapGlobToRegExp(d).test(importSpecifier)` with both sides equal to `lodash`: `startsWith` is true, and the anchored glob cannot match the bare package name, so `every` returns true and the package is treated as not allowed.

Worth splitting the clause by constraint type, so an `allowedExternalImports` subpath entry is described as not whitelisting the package rather than as applying only to direct imports.

**Review by AgentEnder — APPROVED** [review]
The nested violation now names the offending package and reports once per child project, which closes what I raised last round. Two banned packages reached through one child give two distinct messages, and duplicate edges or several resolved versions of one package collapse to a single report.

Approving. One non-blocking follow-up inline: the subpath note I asked for holds for `bannedExternalImports`, but not for `allowedExternalImports`.
